dev-docs/RFCs/v6.2/mapbox-custom-layer-rfc.md

```diff
@@ -0,0 +1,191 @@
+
+### Proposal 1: Callbacks
+
+```js
+import Deck, {ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+const map = new mapboxgl.Map({...});
+
+const deck = new Deck({
+  width: '100%',
+  height: '100%',
+  longitude: -122.4,
+  latitude: 37.8,
+  layers: [
+    new MapboxLayer({
+      id: 'my-scatterplot',
+      insertBefore: 'admin-1-labels',
+      type: ScatterplotLayer,
+      ...
+    })
+  ],
+  onAddLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.addLayer(layer, layer.props.insertBefore)
+    }
+  },
+  onRemoveLayer: (layer) => {
+    if (layer instanceof MapboxLayer) {
+      map.removeLayer(layer.id);
+    }
+  }
+});
+```
+
+Advantages:
+
+* Allows deck.gl and mapbox layers to be ordered (to some degree) within the deck.gl layer list
+* Callbacks are generic and could possibly be useful in other contexts
+
+Downsides:
+
+* Uses `MapboxLayer` inside deck.gl API, ideally that module should be reserved for mapbox API.
+* Relies on callbacks which is not "declarative"
+* Makes it harder to change/control mapbox layer structure, as injection of custom layers is based on deck.gl lifecycle and not directly under application control.
+
+
+
+### Proposal 2: LayerGroups
+
+Ideas:
+
+* "Reuse" the `LayerGroup` concept from the "Layer Group and Operations RFC" to enable the specification of groups of layers that can be rendered "separately".
+* Move towards an official static deck.gl rendering API where the app can add a list of groups and then control how they are rendered.
+
+
+```js
+import Deck, {LayerGroup, ScatterplotLayer} from '@deck.gl/core';
+import {MapboxLayer} from '@deck.gl/mapbox';
+
+class App extends React.Component {
+  render() {
+    return (
+      <Deck
+        width='100%'
+        height='100%'
+        initialViewState={{
+          longitude: -122.4,
+          latitude: 37.8,
+        }}
+
+        // A simple, callbackless way to make deck initialize with the gl context from a sub-component is envisioned...
+        gl='map'
+
+        layers={[
+          // Layer groups create named groups of layers that don't render automatically
+          // In this case we rely on MapboxLayers to render them.
+          new LayerGroup({
+            id: 'below-labels',
+            layers: [
+              new ScatterplotLayer({
+                id: 'my-scatterplot',
+                insertBefore: 'admin-1-labels',
+                type: ScatterplotLayer,
+                ...
+              })
+            ]
+          }),
+
+          // Any layers not in group render automatically as before.
+          new ArcLayer({...});
+        ]}
+        >
+
+        <StaticMap
+          id='map'
+          style="mapbox://styles/mapbox/light-v9"
+          onLoad={this._onLoad.bind(this)/>
+
+      </DeckGL>
+    );
+  }
+
+  _onLoad(map) {
+    const insertBefore = getFirstTextLayerId(map.getStyle());
+    map.addLayer(new MapboxLayer({groupId: })', insertBefore);
```

Comment:
This solution relies on the `onLoad` callback to add layers to mapbox. The biggest issue for kepler.gl, if it uses the "mapbox add-on" API, will be managing the adding/removing/updating of layers in a reactive manner. It is unclear to me how this solution will track layer matching results and add/remove them dynamically from mapbox.

Reply:
> It is unclear to me how this solution will track layer matching results and add/remove them dynamically from mapbox.

Well the main idea is that we use the custom layer API to let the app specify when a deck.gl "layer group" should be rendered in the mapbox render order.  
* The coupling is loose, the custom mapbox style does not contain any layers, it just contains a "layer group" id.
* Mapbox goes through its layers, and if it finds a deck.gl layer group id, asks deck.gl to render that layer group.
* If a layer group with that id is in the deck.gl stack, it gets rendered. If not, it doesn't.
* Both the deck and mapbox layers stacks are manually controlled by the app, so in contrast to the other proposal, there is no issues with mapbox layer stack being autoupdated in callbacks during layer matching.

Reply:
PS - I chose layer groups since I saw them coming in the layer operations RFC, but the same concept could be implemented more simply using a prop that marks a layer as "non-auto rendering". 

```js
  layers: [
     new ScatterplotLayer({
       id: ...,
       autoRender: false,
    })
   ]
```

In this case, that layer would still be matched etc, but an explicit `deck.renderLayer(layerId)` call would be needed to actually draw it.

In some sense, I believe this idea is common to both proposals, and I suspect there are some possible variations on this theme in terms of props etc.

Reply:
Merely skipping render is not enough. The mapbox layers' lifecycles must be managed by another `LayerManager`, as their resources need to be created on a different WebGL context.

Reply:
> there is no issues with mapbox layer stack being autoupdated in callbacks during layer matching.

I may be missing the point here. kepler.gl wants to be able to manage the mapbox layer stack the same way it manages the deck.gl layer stack (i.e. add/remove/update by passsing a new `layers` array instead of doing diffing itself).